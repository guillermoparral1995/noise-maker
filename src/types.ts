import type { ActionTypes as EnvelopeActionTypes } from './components/Controls/EnvelopeControls/store/actions';
import type { ActionTypes as GeneralControlsActionTypes } from './components/Controls/GeneralControls/store/actions';
import type { ActionTypes as FilterActionTypes } from './components/Controls/FilterControls/store/actions';
import type { ActionTypes as LFOActionTypes } from './components/Controls/LFOControls/store/actions';
import { LFOTargetAction } from './providers/AudioContextProvider';

export type ActionTypes =
  | EnvelopeActionTypes
  | GeneralControlsActionTypes
  | FilterActionTypes
  | LFOTargetAction
  | LFOActionTypes;

export enum Waveform {
  SINE = 'sine',
  SQUARE = 'square',
  SAWTOOTH = 'sawtooth',
  TRIANGLE = 'triangle',
}

export enum FilterType {
  LOWPASS = 'lowpass',
  HIGHPASS = 'highpass',
}

export enum Selectors {
  WAVEFORM = 'waveform',
  FILTER = 'filter',
  LFO_TARGET = 'lfoTarget',
  LFO_WAVEFORM = 'lfoWaveform',
}

export enum Knobs {
  VOLUME = 'VOLUME',
  PAN = 'PAN',
  DETUNE = 'DETUNE',
  ATTACK = 'ATTACK',
  DECAY = 'DECAY',
  SUSTAIN = 'SUSTAIN',
  RELEASE = 'RELEASE',
  FILTER_CUTOFF = 'FILTER_CUTOFF',
  FILTER_RESONANCE = 'FILTER_RESONANCE',
  LFO_FREQUENCY = 'LFO_FREQUENCY',
  LFO_AMPLITUDE = 'LFO_AMPLITUDE',
}

export type LFOTarget =
  | Exclude<
      Knobs,
      | Knobs.LFO_FREQUENCY
      | Knobs.LFO_AMPLITUDE
      | Knobs.ATTACK
      | Knobs.DECAY
      | Knobs.SUSTAIN
      | Knobs.RELEASE
    >
  | 'off';

export interface State {
  volume: number;
  pan: number;
  waveform: Waveform;
  attack: number;
  decay: number;
  sustain: number;
  release: number;
  filter: {
    type: FilterType;
    cutoff: number;
    resonance: number;
  };
  lfo: {
    target: LFOTarget;
    waveform: Waveform;
    frequency: number;
    amplitude: number;
  };
}

export enum Actions {
  UPDATE_VOLUME,
  UPDATE_PAN,
  UPDATE_WAVEFORM,
  UPDATE_DETUNE,
  UPDATE_ATTACK,
  UPDATE_DECAY,
  UPDATE_SUSTAIN,
  UPDATE_RELEASE,
  UPDATE_FILTER_TYPE,
  UPDATE_FILTER_CUTOFF,
  UPDATE_FILTER_RESONANCE,
  UPDATE_LFO_TARGET,
  UPDATE_LFO_WAVEFORM,
  UPDATE_LFO_FREQUENCY,
  UPDATE_LFO_AMPLITUDE,
}
export type ActionBuilder<T> = (payload: T) => ActionTypes;
