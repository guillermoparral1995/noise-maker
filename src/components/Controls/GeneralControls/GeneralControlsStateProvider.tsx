import React, { PropsWithChildren, useReducer } from 'react';
import { ActionTypes } from './store/actions';
import initialState, { GeneralControlsState } from './store/initialState';
import reducer from './store/reducer';

interface GeneralControlsStateProviderValue {
  state: GeneralControlsState;
  dispatch: React.Dispatch<ActionTypes>;
}

export const generalControlsStateContext =
  React.createContext<GeneralControlsStateProviderValue>(undefined);

export const GeneralControlsStateProvider = ({
  children,
  __mockDispatch,
}: { __mockDispatch?: React.Dispatch<ActionTypes> } & PropsWithChildren) => {
  const [state, dispatch] = useReducer(reducer, initialState);

  return (
    <generalControlsStateContext.Provider
      value={{ state, dispatch: __mockDispatch ?? dispatch }}
    >
      {children}
    </generalControlsStateContext.Provider>
  );
};
