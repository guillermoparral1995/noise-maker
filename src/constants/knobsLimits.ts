import { Knobs } from '../types';

interface KnobLimit {
  min: number;
  max: number;
  default: number;
}

export const knobsLimits: Record<Knobs, KnobLimit> = {
  [Knobs.VOLUME]: {
    min: 0,
    max: 1,
    default: 0.5,
  },
  [Knobs.PAN]: {
    min: -1,
    max: 1,
    default: 0,
  },
  [Knobs.ATTACK]: {
    min: 0,
    max: 5,
    default: 0,
  },
  [Knobs.DECAY]: {
    min: 0,
    max: 5,
    default: 0,
  },
  [Knobs.SUSTAIN]: {
    min: 0,
    max: 1,
    default: 1,
  },
  [Knobs.RELEASE]: {
    min: 0,
    max: 5,
    default: 0,
  },
  [Knobs.FILTER_CUTOFF]: {
    min: 10,
    max: 5000,
    default: 350,
  },
  [Knobs.FILTER_RESONANCE]: {
    min: 0.0001,
    max: 50,
    default: 1,
  },
  [Knobs.LFO_FREQUENCY]: {
    min: 0.2,
    max: 10,
    default: 1,
  },
  [Knobs.LFO_AMPLITUDE]: {
    min: 0,
    max: 1,
    default: 0.5,
  },
};
