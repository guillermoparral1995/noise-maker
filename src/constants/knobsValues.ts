import {
  updateAttack,
  updateDecay,
  updateDetune,
  updatePitchbend,
  updateRelease,
  updateSustain,
} from '../components/Controls/EnvelopeControls/store/actions';
import {
  updateFilterCutoff,
  updateFilterResonance,
} from '../components/Controls/FilterControls/store/actions';
import {
  updatePan,
  updateVolume,
} from '../components/Controls/GeneralControls/store/actions';
import {
  updateLFO1Amplitude,
  updateLFO1Frequency,
  updateLFO2Amplitude,
  updateLFO2Frequency,
} from '../components/Controls/LFOControls/store/actions';
import { ActionBuilder, Knobs } from '../types';

interface KnobValue {
  label: string;
  min: number;
  max: number;
  default: number;
  midiControl?: number;
  action: ActionBuilder<number>;
  step?: number;
}

export const knobsValues: Record<Knobs, KnobValue> = {
  [Knobs.VOLUME]: {
    label: 'Volume',
    min: 0,
    max: 1,
    default: 0.5,
    midiControl: 74,
    action: updateVolume,
  },
  [Knobs.PAN]: {
    label: 'Pan',
    min: -1,
    max: 1,
    default: 0,
    midiControl: 18,
    action: updatePan,
  },
  [Knobs.DETUNE]: {
    label: 'Detune',
    min: -1200,
    max: 1200,
    default: 0,
    midiControl: 75,
    action: updateDetune,
    step: 1,
  },
  [Knobs.PITCHBEND]: {
    label: 'Pitchbend',
    min: -300,
    max: 300,
    default: 0,
    action: updatePitchbend,
    step: 1,
  },
  [Knobs.ATTACK]: {
    label: 'Attack',
    min: 0,
    max: 5,
    default: 0,
    midiControl: 93,
    action: updateAttack,
  },
  [Knobs.DECAY]: {
    label: 'Decay',
    min: 0,
    max: 5,
    default: 0,
    midiControl: 91,
    action: updateDecay,
  },
  [Knobs.SUSTAIN]: {
    label: 'Sustain',
    min: 0,
    max: 1,
    default: 1,
    midiControl: 73,
    action: updateSustain,
  },
  [Knobs.RELEASE]: {
    label: 'Release',
    min: 0,
    max: 5,
    default: 0,
    midiControl: 79,
    action: updateRelease,
  },
  [Knobs.FILTER_CUTOFF]: {
    label: 'Cut-off',
    min: 10,
    max: 5000,
    default: 350,
    midiControl: 71,
    action: updateFilterCutoff,
    step: 1,
  },
  [Knobs.FILTER_RESONANCE]: {
    label: 'Resonance',
    min: 0.0001,
    max: 50,
    default: 1,
    midiControl: 19,
    action: updateFilterResonance,
    step: 0.1,
  },
  [Knobs.LFO_1_FREQUENCY]: {
    label: 'Frequency',
    min: 0.2,
    max: 10,
    default: 1,
    midiControl: 76,
    action: updateLFO1Frequency,
  },
  [Knobs.LFO_1_AMPLITUDE]: {
    label: 'Amplitude',
    min: 0,
    max: 1,
    default: 0.5,
    midiControl: 16,
    action: updateLFO1Amplitude,
  },
  [Knobs.LFO_2_FREQUENCY]: {
    label: 'Frequency',
    min: 0.2,
    max: 10,
    default: 1,
    midiControl: 77,
    action: updateLFO2Frequency,
  },
  [Knobs.LFO_2_AMPLITUDE]: {
    label: 'Amplitude',
    min: 0,
    max: 1,
    default: 0.5,
    midiControl: 17,
    action: updateLFO2Amplitude,
  },
};
