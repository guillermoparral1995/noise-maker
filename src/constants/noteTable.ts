export interface NoteValue {
  frequency: number;
  mapping: string;
}

type NoteTable = Record<Notes, NoteValue>;

export enum Notes {
  C3 = 'C3',
  CSharp3 = 'C#3',
  D3 = 'D3',
  DSharp3 = 'D#3',
  E3 = 'E3',
  F3 = 'F3',
  FSharp3 = 'F#3',
  G3 = 'G3',
  GSharp3 = 'G#3',
  A3 = 'A3',
  ASharp3 = 'A#3',
  B3 = 'B3',
  C4 = 'C4',
  CSharp4 = 'C#4',
  D4 = 'D4',
  DSharp4 = 'D#4',
  E4 = 'E4',
  F4 = 'F4',
  FSharp4 = 'F#4',
  G4 = 'G4',
  GSharp4 = 'G#4',
  A4 = 'A4',
  ASharp4 = 'A#4',
  B4 = 'B4',
  C5 = 'C5',
}

const noteTable: NoteTable = {
  [Notes.C3]: {
    frequency: 130.81,
    mapping: 'q',
  },
  [Notes.CSharp3]: {
    frequency: 138.59,
    mapping: '2',
  },
  [Notes.D3]: {
    frequency: 146.83,
    mapping: 'w',
  },
  [Notes.DSharp3]: {
    frequency: 155.56,
    mapping: '3',
  },
  [Notes.E3]: {
    frequency: 164.81,
    mapping: 'e',
  },
  [Notes.F3]: {
    frequency: 174.61,
    mapping: 'r',
  },
  [Notes.FSharp3]: {
    frequency: 185,
    mapping: '5',
  },
  [Notes.G3]: {
    frequency: 196,
    mapping: 't',
  },
  [Notes.GSharp3]: {
    frequency: 207.65,
    mapping: '6',
  },
  [Notes.A3]: {
    frequency: 220,
    mapping: 'y',
  },
  [Notes.ASharp3]: {
    frequency: 233.08,
    mapping: '7',
  },
  [Notes.B3]: {
    frequency: 246.94,
    mapping: 'u',
  },
  [Notes.C4]: {
    frequency: 261.63,
    mapping: 'z',
  },
  [Notes.CSharp4]: {
    frequency: 277.18,
    mapping: 's',
  },
  [Notes.D4]: {
    frequency: 293.66,
    mapping: 'x',
  },
  [Notes.DSharp4]: {
    frequency: 311.13,
    mapping: 'd',
  },
  [Notes.E4]: {
    frequency: 329.63,
    mapping: 'c',
  },
  [Notes.F4]: {
    frequency: 349.23,
    mapping: 'v',
  },
  [Notes.FSharp4]: {
    frequency: 369.99,
    mapping: 'g',
  },
  [Notes.G4]: {
    frequency: 392,
    mapping: 'b',
  },
  [Notes.GSharp4]: {
    frequency: 415.3,
    mapping: 'h',
  },
  [Notes.A4]: {
    frequency: 440,
    mapping: 'n',
  },
  [Notes.ASharp4]: {
    frequency: 466.16,
    mapping: 'j',
  },
  [Notes.B4]: {
    frequency: 493.88,
    mapping: 'm',
  },
  [Notes.C5]: {
    frequency: 523.25,
    mapping: ',',
  },
};

export default noteTable;
