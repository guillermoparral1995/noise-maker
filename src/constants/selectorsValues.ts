import {
  updateLFO1Target,
  updateLFO2Target,
} from '@providers/AudioContextProvider/store/actions';
import { updateMIDIInput } from '@providers/MIDIProvider/store/actions';
import {
  ActionBuilder,
  FilterType,
  Knobs,
  LFO1Target,
  LFO2Target,
  Selectors,
  Waveform,
} from '@types';
import { updateWaveform } from '../components/Controls/EnvelopeControls/store/actions';
import { updateFilterType } from '../components/Controls/FilterControls/store/actions';
import {
  updateLFO1Waveform,
  updateLFO2Waveform,
} from '../components/Controls/LFOControls/store/actions';

type DropdownValue = Waveform | FilterType | LFO1Target | LFO2Target | string;

export interface DropdownOption {
  label: string;
  value: DropdownValue;
}

interface SelectorValues {
  label: string;
  action: ActionBuilder<DropdownValue>;
  options: DropdownOption[];
}

const waveforms: DropdownOption[] = [
  {
    label: 'Sine',
    value: Waveform.SINE,
  },
  {
    label: 'Square',
    value: Waveform.SQUARE,
  },
  {
    label: 'Sawtooth',
    value: Waveform.SAWTOOTH,
  },
  {
    label: 'Triangle',
    value: Waveform.TRIANGLE,
  },
];

const filterTypes: DropdownOption[] = [
  {
    label: 'Low-pass',
    value: FilterType.LOWPASS,
  },
  {
    label: 'High-pass',
    value: FilterType.HIGHPASS,
  },
];

const lfo1Targets: DropdownOption[] = [
  {
    label: 'Off',
    value: 'off',
  },
  {
    label: 'Volume',
    value: Knobs.VOLUME,
  },
  {
    label: 'Pan',
    value: Knobs.PAN,
  },
  {
    label: 'Cut-off',
    value: Knobs.FILTER_CUTOFF,
  },
  {
    label: 'Resonance',
    value: Knobs.FILTER_RESONANCE,
  },
  {
    label: 'Detune',
    value: Knobs.DETUNE,
  },
];

export const selectorValues: Record<Selectors, SelectorValues> = {
  [Selectors.WAVEFORM]: {
    label: 'Waveform',
    action: updateWaveform,
    options: waveforms,
  },
  [Selectors.LFO_1_WAVEFORM]: {
    label: 'Waveform',
    action: updateLFO1Waveform,
    options: waveforms,
  },
  [Selectors.LFO_2_WAVEFORM]: {
    label: 'Waveform',
    action: updateLFO2Waveform,
    options: waveforms,
  },
  [Selectors.FILTER]: {
    label: 'Type',
    action: updateFilterType,
    options: filterTypes,
  },
  [Selectors.LFO_1_TARGET]: {
    label: 'Target',
    action: updateLFO1Target,
    options: lfo1Targets,
  },
  [Selectors.LFO_2_TARGET]: {
    label: 'Target',
    action: updateLFO2Target,
    options: [
      ...lfo1Targets,
      {
        label: 'LFO 1 Frequency',
        value: Knobs.LFO_1_FREQUENCY,
      },
      {
        label: 'LFO 1 Amplitude',
        value: Knobs.LFO_1_AMPLITUDE,
      },
    ],
  },
  [Selectors.MIDI_INPUT]: {
    label: 'MIDI Input',
    action: updateMIDIInput,
    options: [],
  },
};
