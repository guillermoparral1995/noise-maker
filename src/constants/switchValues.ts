import { switchDelayTrails } from '../components/Controls/DelayControls/store/actions';
import { switchDelay } from '../providers/AudioContextProvider/store/actions';
import { Switchs } from '../types';

export const switchValues = {
  [Switchs.DELAY]: {
    label: 'Delay',
    action: switchDelay,
  },

  [Switchs.DELAY_TRAILS]: {
    label: 'Trails',
    action: switchDelayTrails,
  },
};
