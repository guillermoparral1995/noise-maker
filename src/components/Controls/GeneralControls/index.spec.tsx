import { cleanup, render } from '@testing-library/react';
import React from 'react';
import { GeneralControls_ as GeneralControls } from '.';
import {
  AudioContextMock,
  GainNodeMock,
  StereoPannerNodeMock,
} from '../../../../__mocks__';
import { knobsValues } from '../../../constants/knobsValues';
import {
  AudioContextProvider,
  LFOMock,
} from '../../../providers/AudioContextProvider';
import {
  withMockedMIDIInput,
  withMockedMIDINoInput,
} from '../../../providers/MIDIProvider';
import { Actions, Knobs, LFO1Target, LFO2Target } from '../../../types';
import { GeneralControlsStateProvider } from './GeneralControlsStateProvider';

const __mockGetInputByName = jest.fn();
jest.mock('webmidi', () => {
  return {
    ...jest.requireActual('webmidi'),
    WebMidi: {
      inputs: [],
      enable: jest.fn(),
      getInputByName: () => __mockGetInputByName(),
    },
  };
});

describe('General controls', () => {
  const mockContext = new AudioContextMock();
  const mockVolume = new GainNodeMock();
  const mockPan = new StereoPannerNodeMock();
  const mockLfo1: LFOMock<LFO1Target> = {
    target: 'off',
    output: new GainNodeMock(),
  };
  const mockLfo2: LFOMock<LFO2Target> = {
    target: 'off',
    output: new GainNodeMock(),
  };

  afterEach(() => {
    cleanup();
  });

  it('should render correctly', async () => {
    const { findByTestId } = render(
      withMockedMIDINoInput(
        <AudioContextProvider
          __mocks={{
            context: mockContext,
            volume: mockVolume,
            pan: mockPan,
            lfo1: mockLfo1,
            lfo2: mockLfo2,
          }}
        >
          <GeneralControlsStateProvider>
            <div data-testid="general-controls">
              <GeneralControls></GeneralControls>
            </div>
          </GeneralControlsStateProvider>
        </AudioContextProvider>,
      ),
    );

    const controls = await findByTestId('general-controls');
    expect(controls).toMatchSnapshot();
  });

  it('should trigger dispatch on MIDI knob change', async () => {
    const mockDispatch = jest.fn();
    let mockEventName: string;
    const mockAddListener = jest.fn(
      (eventName: string, listener: (e: object) => void) => {
        mockEventName = eventName;
        listener({
          value: 0.1,
          controller: { number: knobsValues[Knobs.VOLUME].midiControl },
        });
      },
    );

    __mockGetInputByName.mockReturnValue({
      addListener: mockAddListener,
    });

    const { findByTestId } = render(
      withMockedMIDIInput(
        <AudioContextProvider
          __mocks={{
            context: mockContext,
            volume: mockVolume,
            pan: mockPan,
            lfo1: mockLfo1,
            lfo2: mockLfo2,
          }}
        >
          <GeneralControlsStateProvider __mockDispatch={mockDispatch}>
            <GeneralControls></GeneralControls>
          </GeneralControlsStateProvider>
        </AudioContextProvider>,
      ),
    );

    await findByTestId(Knobs.VOLUME);
    expect(mockEventName).toEqual('controlchange');
    expect(mockDispatch).toHaveBeenCalledWith({
      type: Actions.UPDATE_VOLUME,
      payload: 0.1,
    });
  });

  it('should connect lfo1 with param', async () => {
    const mockLfo1: LFOMock<LFO1Target> = {
      target: Knobs.VOLUME,
      output: new GainNodeMock(),
    };
    const mockDispatch = jest.fn();
    const { findByTestId } = render(
      withMockedMIDIInput(
        <AudioContextProvider
          __mocks={{
            context: mockContext,
            volume: mockVolume,
            pan: mockPan,
            lfo1: mockLfo1,
            lfo2: mockLfo2,
          }}
        >
          <GeneralControlsStateProvider __mockDispatch={mockDispatch}>
            <GeneralControls></GeneralControls>
          </GeneralControlsStateProvider>
        </AudioContextProvider>,
      ),
    );
    await findByTestId(Knobs.VOLUME);

    expect(mockLfo1.output.__mockConnect).toHaveBeenCalledWith(mockVolume.gain);
    expect(mockLfo2.output.__mockConnect).not.toHaveBeenCalled();
  });

  it('should connect lfo2 with param', async () => {
    const mockLfo2: LFOMock<LFO2Target> = {
      target: Knobs.PAN,
      output: new GainNodeMock(),
    };
    const mockDispatch = jest.fn();
    const { findByTestId } = render(
      withMockedMIDIInput(
        <AudioContextProvider
          __mocks={{
            context: mockContext,
            volume: mockVolume,
            pan: mockPan,
            lfo1: mockLfo1,
            lfo2: mockLfo2,
          }}
        >
          <GeneralControlsStateProvider __mockDispatch={mockDispatch}>
            <GeneralControls></GeneralControls>
          </GeneralControlsStateProvider>
        </AudioContextProvider>,
      ),
    );
    await findByTestId(Knobs.VOLUME);

    expect(mockLfo2.output.__mockConnect).toHaveBeenCalledWith(mockPan.pan);
    expect(mockLfo1.output.__mockConnect).not.toHaveBeenCalled();
  });
});
