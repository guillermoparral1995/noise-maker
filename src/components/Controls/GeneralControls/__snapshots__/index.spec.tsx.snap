// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`General controls should render correctly 1`] = `
<div
  data-testid="general-controls"
>
  <h3>
    Master
  </h3>
  <div
    data-testid="VOLUME"
    tabindex="0"
  >
    <label
      class="label"
      id="knob_label_VOLUME"
    >
      Volume
    </label>
    <div
      class="slider"
    >
      <div
        class="p-slider p-component p-slider-horizontal"
        data-pc-name="slider"
        data-pc-section="root"
        id="VOLUME"
        name="VOLUME"
      >
        <span
          class="p-slider-range"
          data-pc-section="range"
          style="position: absolute; width: 50%;"
        />
        <span
          aria-orientation="horizontal"
          aria-valuemax="1"
          aria-valuemin="0"
          aria-valuenow="50"
          class="p-slider-handle"
          data-pc-section="handle"
          role="slider"
          style="position: absolute; left: 50%;"
          tabindex="0"
        />
      </div>
    </div>
  </div>
  <div
    data-testid="PAN"
    tabindex="0"
  >
    <label
      class="label"
      id="knob_label_PAN"
    >
      Pan
    </label>
    <div
      class="slider"
    >
      <div
        class="p-slider p-component p-slider-horizontal"
        data-pc-name="slider"
        data-pc-section="root"
        id="PAN"
        name="PAN"
      >
        <span
          class="p-slider-range"
          data-pc-section="range"
          style="position: absolute; width: 0%;"
        />
        <span
          aria-orientation="horizontal"
          aria-valuemax="1"
          aria-valuemin="-1"
          aria-valuenow="0"
          class="p-slider-handle"
          data-pc-section="handle"
          role="slider"
          style="position: absolute; left: 0%;"
          tabindex="0"
        />
      </div>
    </div>
  </div>
</div>
`;
